### PR TITLE
HSEARCH-3049 Implement the missing work types compared to Search 5 (except DeleteByQuery)

### PR DESCRIPTION
**Pull request by yrodiere**
https://hibernate.atlassian.net//browse/HSEARCH-3049

*Before* you have a look at the commits, I would recommend having a look at the "Explicit indexing" section of the documentation that I just added. Hopefully it will explain why those changes are being made clearly enough :)

### CODE REVIEW

File: documentation/src/main/asciidoc/mapper-orm-indexing-explicit.asciidoc [hunk 1]
```diff
@@ -0,0 +1,374 @@
...
+or simply restoring a database backup.
+* when indexes had to be wiped,
+for example because the Hibernate Search mapping or some core settings changed.
+* when automatic indexing had to be disabled for performance reasons,
+and periodic reindexing (every night, ...) is preferred.
+
+[NOTE]
+====
+When the the entity mapping to the index changes,
+this rule of thumb can help to determine if the changes require a full reindexing or not:
```
Comment (fax4ever) [inline documentation/src/main/asciidoc/mapper-orm-indexing-explicit.asciidoc]: I don't think that those two lines be necessary.

File: documentation/src/main/asciidoc/mapper-orm-indexing-explicit.asciidoc [hunk 2]
```diff
@@ -0,0 +1,374 @@
...
+
+* Wiping the index and reindexing is *not* required if the mapping changes include *only*:
```
Comment (fax4ever) [inline documentation/src/main/asciidoc/mapper-orm-indexing-explicit.asciidoc]: I would remove the external list, keeping the internal one.

File: documentation/src/main/asciidoc/mapper-orm-indexing-explicit.asciidoc [hunk 3]
```diff
@@ -0,0 +1,374 @@
...
+   e.g. removing an index field, or removing the need for a field to be stored.
+* Wiping the index and reindexing *is* required if the mapping changes include anything that is not mentioned above.
+====
+
+To address these use cases, Hibernate Search exposes several APIs
+explained if the following sections.
+
+As with everything in Hibernate Search,
+these APIs only affect the Hibernate Search indexes:
+they do not write to the database.
```
Comment (fax4ever) [inline documentation/src/main/asciidoc/mapper-orm-indexing-explicit.asciidoc]: Maybe /write/write anything/s ?
Or they do not touch the database ?

File: documentation/src/main/asciidoc/mapper-orm-indexing-explicit.asciidoc [hunk 4]
```diff
@@ -0,0 +1,374 @@
...
+<5> After the loop, commit the transaction.
+The changes stored in the internal buffer, as well as the remaining entities that were not flushed/cleared,
+will be indexed automatically.
+====
+
+[NOTE]
+====
+The multi-transaction solution and the periodic `process()` solution can be combined,
+breaking down the process in multiple medium-sized transactions,
+and periodically calling `flush`/`process`/`clear` inside each transaction.
```
Comment (fax4ever) [inline documentation/src/main/asciidoc/mapper-orm-indexing-explicit.asciidoc]: Maybe we can add that this last combined strategy is the more flexible and the more suitable for a fine tuning?

File: documentation/src/main/asciidoc/mapper-orm-indexing-explicit.asciidoc [hunk 5]
```diff
@@ -0,0 +1,374 @@
...
+The operation won't be executed immediately,
+but will be delayed until the transaction is committed.
+<5> Commit the transaction, allowing Hibernate Search to actually write the document to the index.
+====
+
+.Explicitly deleting an entity from the index using `SearchSessionWritePlan`
+====
+[source, JAVA, indent=0]
+----
+include::{sourcedir}/org/hibernate/search/documentation/mapper/orm/indexing/HibernateOrmExplicitIndexingIT.java[tags=write-plan-addOrUpdate]
```
Comment (fax4ever) [inline documentation/src/main/asciidoc/mapper-orm-indexing-explicit.asciidoc]: I think there's a wrong tag here.

File: documentation/src/main/asciidoc/mapper-orm-indexing-explicit.asciidoc [hunk 6]
```diff
@@ -0,0 +1,374 @@
...
+
+This interface offers the following methods:
+
+`purge()`::
+Purge the indexes targeted by this writer, removing all documents.
++
+When using multi-tenancy, only documents of one tenant will be removed:
+the tenant of the session from which this writer originated.
+`purgeAsync()`::
+Asynchronous version of `purge()`.
```
Comment (fax4ever) [inline documentation/src/main/asciidoc/mapper-orm-indexing-explicit.asciidoc]: I would add `returning a CompletableFuture`.

File: documentation/src/main/asciidoc/mapper-orm-indexing-explicit.asciidoc [hunk 7]
```diff
@@ -0,0 +1,374 @@
...
+`flush()`::
+Flush to disk the changes to indexes that were not committed yet.
+In the case of backends with a transaction log (Elasticsearch),
+also apply operations from the transaction log that were not applied yet.
++
+This is generally not useful as Hibernate Search commits changes automatically.
+Only to be used by experts fully aware of the implications.
+`flushAsync()`::
+Asynchronous version of `flush()`.
```
Comment (fax4ever) [inline documentation/src/main/asciidoc/mapper-orm-indexing-explicit.asciidoc]: same here

File: documentation/src/main/asciidoc/mapper-orm-indexing-explicit.asciidoc [hunk 8]
```diff
@@ -0,0 +1,374 @@
...
+`optimize()`::
+Merge all segments of the indexes targeted by this writer into a single one.
++
+Note this operation may affect performance positively as well as negatively.
+As a rule of thumb, if indexes are read-only for extended periods of time,
+then calling `optimize()` may improve performance.
+If indexes are written to, then calling `optimize()`
+is likely to degrade read/write performance overall.
+`optimizeAsync()`::
+Asynchronous version of `optimize()`.
```
Comment (fax4ever) [inline documentation/src/main/asciidoc/mapper-orm-indexing-explicit.asciidoc]: same here

File: mapper/orm/src/main/java/org/hibernate/search/mapper/orm/session/SearchSessionWritePlan.java [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later
+ * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ */
+package org.hibernate.search.mapper.orm.session;
+
+/**
+ * An interface for writing to indexes in the context of an ORM Session.
+ * <p>
+ * When a transaction is active, operations are queued internally,
```
[the remark this answers is not in the document]
Comment (yrodiere) [reply mapper/orm/src/main/java/org/hibernate/search/mapper/orm/session/SearchSessionWritePlan.java]: At the moment, I don't know. This is a different ticket: https://hibernate.atlassian.net/browse/HSEARCH-3069.

I'll add a TODO here.

File: integrationtest/mapper/orm/src/test/java/org/hibernate/search/integrationtest/mapper/orm/session/SearchSessionWritePlanIT.java [hunk 10]
```diff
@@ -0,0 +1,390 @@
...
+			writePlan.purge( IndexedEntity1.class, 44 );
+			writePlan.purge( IndexedEntity1.class, 44 );
+
+			writePlan.addOrUpdate( entity5 );
+			writePlan.purge( IndexedEntity1.class, 5 );
+
+			writePlan.purge( IndexedEntity1.class, 6 );
+			writePlan.addOrUpdate( entity6 );
+
+			backend1Mock.expectWorks( IndexedEntity1.INDEX_NAME )
```
[the remark this answers is not in the document]
Comment (yrodiere) [reply integrationtest/mapper/orm/src/test/java/org/hibernate/search/integrationtest/mapper/orm/session/SearchSessionWritePlanIT.java]: I forgot. Done.

File: mapper/pojo/src/main/java/org/hibernate/search/mapper/pojo/logging/impl/Log.java [hunk 11]
```diff
@@ -408,4 +413,11 @@ SearchException cannotResolveContainerExtractorName(String extractorName,
 	SearchException invalidAnnotationTypeForBuilder(Object builder,
 			@FormatWith(ClassFormatter.class) Class<?> annotationType);
 
+	@Message(id = ID_OFFSET_2 + 55,
+			value = "Type '%1$s' is contained in an indexed type but is not itself indexed,"
+					+ " thus entity with identifier '%2$s' cannot be purged."
+					+ " Use delete() and pass the entity instead of just the identifier."
+	)
+	SearchException cannotPurgeNonIndexedContainedType(
```
[the remark this answers is not in the document]
Comment (yrodiere) [reply mapper/pojo/src/main/java/org/hibernate/search/mapper/pojo/logging/impl/Log.java]: We didn't. Done.

File: documentation/src/test/java/org/hibernate/search/documentation/mapper/orm/indexing/HibernateOrmExplicitIndexingIT.java [hunk 12]
```diff
@@ -0,0 +1,341 @@
...
+			}
+			// end::persist-automatic-indexing-periodic-execute[]
+
+			assertBookCount( entityManager, NUMBER_OF_BOOKS );
+		} );
+	}
+
+	@Test
+	public void persist_automaticIndexing_periodicProcess_multipleTransactions() {
+		EntityManagerFactory entityManagerFactory = setup( HibernateOrmAutomaticIndexingStrategyName.SESSION );
```
[the remark this answers is not in the document]
Comment (yrodiere) [reply documentation/src/test/java/org/hibernate/search/documentation/mapper/orm/indexing/HibernateOrmExplicitIndexingIT.java]: Neither one, actually. Fixed.

**Comment by yrodiere** [comment]
I addressed most of your comments. Thanks. Waiting for CI and then we can merge.

The only things I did not address were the remarks about the names. I agree they are not great, but there are a few constraints to keep in mind. I will open a ticket so that we can sleep on it and revisit the names before we go Beta.

**Comment by fax4ever** [comment]
Merged thanks!

**Pull request merged**